### ISSUE TITLE
A note on numerical preconditions in MaxPool gradient translation-invariance testing

### ISSUE THREAD

**Issue by XiaoHuaOvo-tech** [tensorflow/tensorflow#107543]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

TF2.10.0

### Custom code

Yes

### OS platform and distribution

Liunx Ubuntu 22.04

### CUDA/cuDNN version

cuDNN 8.6

### GPU model and memory

Tesla V100

### Current behavior?

Hello TensorFlow team,

We’d like to report and clarify an observation from metamorphic testing of MaxPool2D gradients, specifically about the numerical preconditions of a translation-invariance metamorphic relation (MR).

Background (the MR we tested)

We tested the following relation for max pooling:

Constant shift: ( X' = X + c )

Expectation (when argmax routing is preserved): ( gX' \approx gX )

What we observed

With padding = SAME and stride = 1, we observe cases where this MR is violated with a very large violation metric (vR), and the behavior is reproducible on both CPU and GPU.

In particular, when the shift magnitude (|c|) is large (e.g., hundreds or larger), gradients w.r.t. the input can change significantly even though the forward operation is a constant shift.

Our analysis / hypothesis

We believe this is not an implementation bug, but a consequence of finite-precision floating-point effects:

For large (c), input values enter a regime where float32 resolution (ULP) becomes coarse relative to small value differences.

Values that were previously strictly ordered can become equal or nearly equal after the shift (implicit ties).

This can change argmax routing in max pooling.

Therefore the backward gradient routing changes, and the translation-invariance MR becomes inapplicable (the “order preserved” precondition no longer holds under finite precision).

Questions 

We’d like to confirm:

Is it an implicit assumption (in design/spec) that transformations should preserve relative ordering under finite precision when reasoning about MaxPool gradients?

For gradient testing/verification, would it be more appropriate to treat such cases as Not Applicable (N/A) rather than incorrect behavior?

Would it be helpful to add a brief documentation note that large constant shifts may affect MaxPool gradient routing due to finite-precision / tie effects?

Reproduction

We have a minimal reproducible script and concrete configurations (TF version, CPU/GPU) demonstrating this effect, and we can provide it in the thread if needed.

Thank you for your time and for the continued work on TensorFlow.

Best regards,
XiaoHuaOvo

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import numpy as np
import os
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "2"

def vR(a, b):
    a, b = a.ravel(), b.ravel()
    return np.max(np.abs(a - b) / (1e-5 + 1e-3 * np.maximum(np.abs(a), np.abs(b))))

def grad(x):
    with tf.GradientTape() as t:
        t.watch(x)
        y = tf.nn.max_pool2d(tf.transpose(x, [0,2,3,1]),
                             ksize=2, strides=1, padding="SAME")
        y = tf.transpose(y, [0,3,1,2])
    return t.gradient(y, x, output_gradients=tf.ones_like(y))

for dev in ["/cpu:0", "/gpu:0"]:
    if dev == "/gpu:0" and not tf.config.list_physical_devices("GPU"):
        continue
    with tf.device(dev):
        print(f"\n=== {dev} TF {tf.__version__} ===")

        # 非 tie，但非常接近
        x = tf.constant([[[[1.0000000, 1.0000001],
                           [0.0,       0.0      ]]]], tf.float32)

        g0 = grad(x)
        g_small = grad(x + 1.0)
        g_large = grad(x + 1000.0)

        print("small shift vR =", vR(g0.numpy(), g_small.numpy()))
        print("large shift vR =", vR(g0.numpy(), g_large.numpy()))
        print("gX original =", g0.numpy())
        print("gX large    =", g_large.numpy())
```

### Relevant log output

```shell

```

**Comment by XiaoHuaOvo-tech** [comment]
Additional note: even a “small” constant shift (e.g., c=1) can violate the MR if the original max candidates differ by ~1e-7 in float32. This is because float32 ULP increases with magnitude; after shifting, two previously distinct values may round to the same float32 value, creating implicit ties and changing argmax routing.
To make the effect clearer, we can choose a less extreme delta (e.g., 1e-5) so that small shifts preserve ordering but large shifts (e.g., +1000) may collapse ordering into ties.

**Comment by nileshau-afk** [comment]
Hi @XiaoHuaOvo-tech, 

Thank you for your patience, and apologies for the late reply. Your analysis regarding the observed behavior of `MaxPool2D` gradients under large constant shifts is accurate. This is indeed a consequence of finite-precision floating-point arithmetic, where large shifts can alter the relative ordering of values at the ULP level, thereby changing the `argmax` routing and subsequently the gradient path. This is expected numerical behavior rather than an implementation bug.

 If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

We look forward to seeing your contributions!
